pkg/tailscale/net.go

```diff
@@ -0,0 +1,346 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"net/netip"
+	"reflect"
+	"strings"
+
+	jnipkg "github.com/tailscale/tailscale-android/pkg/jni"
+	"github.com/tailscale/wireguard-go/tun"
+	"golang.org/x/sys/unix"
+	"inet.af/netaddr"
+	"tailscale.com/net/dns"
+	"tailscale.com/net/interfaces"
+	"tailscale.com/util/dnsname"
+	"tailscale.com/wgengine/router"
+)
+
+import "C"
+
+// errVPNNotPrepared is used when VPNService.Builder.establish returns
+// null, either because the VPNService is not yet prepared or because
+// VPN status was revoked.
+var errVPNNotPrepared = errors.New("VPN service not prepared or was revoked")
+
+// errMultipleUsers is used when we get a "INTERACT_ACROSS_USERS" error, which
+// happens due to a bug in Android. See:
+//
+//	https://github.com/tailscale/tailscale/issues/2180
+var errMultipleUsers = errors.New("VPN cannot be created on this device due to an Android bug with multiple users")
+
+// Report interfaces in the device in net.Interface format.
+func (a *App) getInterfaces() ([]interfaces.Interface, error) {
+	var ifaceString string
+	err := jnipkg.Do(a.jvm, func(env *jnipkg.Env) error {
+		cls := jnipkg.GetObjectClass(env, a.appCtx)
+		m := jnipkg.GetMethodID(env, cls, "getInterfacesAsString", "()Ljava/lang/String;")
+		n, err := jnipkg.CallObjectMethod(env, a.appCtx, m)
```

Comment:
this string needs to be released once we're done copying it

Reply:
I didn't change this here because we're getting rid of a lot of this in our PR to use gomobile